Test for C++ API compatibility using GCC and MSVC SARIF diagnostics

MisterDA opened the pull request:
Here's a fixed test that includes all OCaml headers in a C⁠+⁠+ file with a simple function that releases and re-acquires the runtime system. The tests parses the output from GCC 13 and MSVC as [Static Analysis Results Interchange Format (SARIF)](https://sarif.info/). Using this format makes it easier to ignore unavoidable errors, such as using flexible arrays in C⁠+⁠+ (this is only supported as an extension). 
I introduce a very basic configure test to check for a C⁠+⁠+ compiler. The `cl.exe` front-end automatically selects the C or C⁠+⁠+ backend based on the file extension, while the `gcc` front-end interprets its input as C⁠+⁠+ with the `-x c++` option. The check also test whether gcc defaults to C⁠+⁠+11 or newer, and is only activated when ocamltest is enabled.
~I've had to hack ocamltest to retry deleting files. On Windows, a file cannot be deleted if it's still opened by a process. Rather than understanding the root cause, I've added countermeasures. I've also hacked ocamltest to write its temporary files in a sub-directory of the temporary directory, as I grew tired of chasing them.~
It would be nicer from a user point-of-view to have access to the usual warning and errors reports of the compiler in addition to the SARIF output, but this is only possible since GCC 15 (not yet available in CI). Note that GCC 15 also ships with [`sarif-replay`](https://manpages.debian.org/experimental/gcc-15-diagnostics/sarif-replay-15.1.en.html), which can replay diagnostics emitted by GCC from a SARIF file. I use it as such:

```console
$ docker build -t sarif-replay -f - <<EOF
# syntax=docker/dockerfile:1
FROM debian:unstable
RUN --mount=type=cache,target=/var/cache/apt,sharing=locked \
    --mount=type=cache,target=/var/lib/apt,sharing=locked \
    apt update && apt-get --no-install-recommends install -y \
    gcc-15-diagnostics
ENTRYPOINT ["/usr/bin/sarif-replay-15"]
EOF
$ docker run --rm --mount type=bind,source=$PWD/stubs.sarif,target=/tmp/stubs.sarif \
             -t sarif-replay /tmp/stubs.sarif
```

In the meantime, the test will diligently log where it's dumped the SARIF output, and display it folded in the GitHub Actions logs.

The https://github.com/MisterDA/ocaml/pull/151 PR shows that this test indeed catches the issues we've had in the past weeks with C⁠+⁠+ compatibility.

gasche commented:
Sorry for the naive question, but can you re-explain why we need SARIF? I understand that flexible array members can create errors different from what you want to test, but would a naive approach work?
- maybe you could filter out this error from the stderr output, and look at the rest?
- don't compilers have flags to tell them to accept flexible array members also in C++ mode, that could be passed to them?

(My understanding is that the shiny SARIF requirement means that you have a specific set of compilers that you support, maybe cl.exe, clang, gcc, and gcc-which-is-in-fact-clang; could one reasonably express this test as a shell-script that checks which of these compilers is present, invokes them directly with the appropriate compiler flags, and filters the output in a compiler-specific way if necessary?)

I also feel vaguely curious of why it is very important that ocamltest reliably removes files in this test. (The test script does not contain explicit removals, which of the actions implies a removal and why cannot removal errors be ignored?)

MisterDA commented:
Thanks for the feedback :)

> Sorry for the naive question, but can you re-explain why we need SARIF? I understand that flexible array members can create errors different from what you want to test, but would a naive approach work?
>
> - maybe you could filter out this error from the stderr output, and look at the rest?

It was my first approach, but it felt very ad-hoc to parse the output of the compiler. I thought that parsing a more-or-less standard format (SARIF is a JSON Schema) was more reliable. I can reject specific errors in 4 lines of jq. Can you imagine what an awk script to do that would look like?

> don't compilers have flags to tell them to accept flexible array members also in C++ mode, that could be passed to them?

Not that I'm aware of. Also sometimes MSVC reports error for invalid C++ constructs that are valid in C when processing C programs, so all of that is quite flaky.

> (My understanding is that the shiny SARIF requirement means that you have a specific set of compilers that you support, maybe cl.exe, clang, gcc, and gcc-which-is-in-fact-clang; could one reasonably express this test as a shell-script that checks which of these compilers is present, invokes them directly with the appropriate compiler flags, and filters the output in a compiler-specific way if necessary?)

The test supports GCC 13 and MSVC. IIRC MSVC supported SARIF before it started supporting C11 atomic ;) The [`deps.sh`](https://github.com/ocaml/ocaml/pull/14421/changes#diff-af5b5022727d129b8d384fbe04d95db62ced57b458f0209c14ffa017c975ec90) script already tests for these two compilers and skips the test otherwise, and the [`cxx-sarif.h`](https://github.com/ocaml/ocaml/pull/14421/changes#diff-411fb74e262efdfa8a267153ac37cb1ab0583afe96ccf147c357640ab3f2ffee) script calls the compiler and parses its SARIF output using jq.

Do you mean looking for other compilers than the one ocaml's been configured to use? My idea is that this test doesn't have to work in all configurations as long as it can be triggered in the CI on Unix and Windows.

> I also feel vaguely curious of why it is very important that ocamltest reliably removes files in this test. (The test script does not contain explicit removals, which of the actions implies a removal and why cannot removal errors be ignored?)

On success produced files are removed. I swear I saw a error related to that. I've reverted the commit, if the error happens again we'll see it.

dra27 commented:
> On success produced files are removed. I swear I saw a error related to that. I've reverted the commit, if the error happens again we'll see it.

The Jenkins workers get this sometimes (although I’ve generally seen that with executables). FWIW, I’d fix that using an approach more like https://github.com/ocaml/ocaml/pull/13921/commits/78fc5774e856cfbe133750c66dfab48e3ab75452 - i.e. loop until a lock can be acquired, then use delete on close to nuke the file. However, I wouldn’t recommend we go down reviewing any of this for a very sporadic failure in the testsuite

dra27 commented on testsuite/tests/lib-unix/unix-sockaddr/sockaddr_cxx.ml:
```diff
@@ -1,11 +1,12 @@
 (* TEST
+ has-cxx;
  not-msvc;
```
Does this have to be skipped on MSVC now?

dra27 commented on testsuite/tests/cxx-api/all_includes.ml:
```diff
@@ -1,13 +1,41 @@
- readonly_files = "all-includes.h";
- not-msvc;
- flags = "-ccopt -x -ccopt c++ -ccopt -std=c++11";
+ has-cxx;
```
Just a note, as I puzzled myself - this test skips on AppVeyor because the testsuite runs bytecode-only (this test could be explicitly done with the dynlink ones, but I don't think it's worth it?)

dra27 commented on testsuite/tests/cxx-api/deps.sh:
```diff
@@ -0,0 +1,32 @@
+set -e
+
+if command -v cygpath >/dev/null 2>&1; then
+  ocamltest_response="$(cygpath -u "${ocamltest_response:?}")"
```
I don't think this is necessary - Windows paths are valid in Cygwin, and we're only using it for piping?

Can you just have `set -eu` above, rather than all the `:?`, or am I missing something?

MisterDA replied:
Yes, that's better, thanks!

dra27 commented on testsuite/tests/cxx-api/deps.sh:
```diff
@@ -0,0 +1,32 @@
+  exit "$TEST_SKIP"
+fi
+
+if [ "${ccomp_type:?}" = "cc" ]; then
```
`case "${ccomp_type}"`?

dra27 commented on aclocal.m4:
```diff
@@ -643,3 +643,23 @@ AC_DEFUN([OCAML_CHECK_WINDOWS_TRIPLET], [
+      [ocaml_cv_prog_cxx="$CC"],
+      [ocaml_cv_prog_cxx=""])
+      CC="$saved_CC"],
+    [ocaml_cv_prog_cxx="$CC"])])
```
I found a bit subtle that this is "cl always C++" - perhaps a comment that MSVC always supports C++ here?

dra27 commented on ocamltest/ocamltest_stdlib.ml:
```diff
@@ -26,6 +26,20 @@ module Char = struct
     c = ' ' || c = '\012' || c = '\n' || c = '\r' || c =  '\t'
 end
 
+module Sys_make_directory = struct
```
Nittily - I think this could be achieved by moving `Filename` below `Sys`? It's a bit weird because that's not the dependency order of the actual Stdlib, but I don't think that matters here?

dra27 commented on ocamltest/ocaml_variables.ml:
```diff
@@ -50,6 +50,12 @@ let cc = make ("cc",
+  "Command to use to invoke the C++ compiler")
+
+let ccomp_type = make ("ccomp_type",
+  "Type of C compiler (msvc, cc, etc.)")
```
```suggestion
  "Type of C compiler (["cl"] for MSVC or ["cc"])")
```

There are predicates testing for MSVC, which makes exposing this as a variable seem a bit strange - but I'm not sure there's a clearer way to communicate it to the shell script?

MisterDA replied:
Yes, I thought exposing `ccomp_type` (its possible values are `msvc` and `cc` something entirely different) was the easiest way of exposing to the test script that msvc-style CLI was used.

dra27 commented on ocamltest/ocamltest_stdlib.ml:
```diff
@@ -41,6 +55,11 @@ module Filename = struct
+  let () =
+    let dirname = Filename.(concat (get_temp_dir_name ()) "ocamltest") in
+    Filename.set_temp_dir_name dirname;
+    Sys_make_directory.make_directory dirname
```
Windows hat on (no tmpfs) - does this ever get cleaned up?

MisterDA replied:
Good catch, I hope an `at_exit` will work. I wonder if `at_exit (Sys.rm_rf dirname)` is a good strategy if there are parallel calls to ocamltest.

dra27 replied:
Ah, I misread it slightly - I think you'd still get the effect if you created an entire directory instead (i.e. `/tmp/ocamltest-234afe`). I was thinking you'd just `rmdir` it - it's an error if there are actually files left in it, I think?

dra27 commented on testsuite/tests/cxx-api/all-includes.h.sh:
```diff
@@ -0,0 +1,11 @@
+  test_build_directory="$(cygpath -u "${test_build_directory:?}")"
+fi
+
+exec find "${ocamlsrcdir:?}"/runtime/caml "$ocamlsrcdir"/otherlibs/*/caml \
```
`find` is a terrifyingly difficult tool to get portably right (`-execdir` is a GNUism, I think?) - I'm tempted rather to generate this header as part of the build, where you could just `$(wildcard` in the build system?

MisterDA replied:
Good idea, and it happened that my `PATH` was corrupted, which `find` likes to complain about.

dra27 commented on testsuite/tests/cxx-api/all_includes.ml:
```diff
@@ -1,13 +1,41 @@
+ include systhreads;
+ script = "sh ${test_source_directory}/deps.sh";
+ script;
+ {
```
It's not instantly clear what the two steps are doing differently - a quick comment here that it's testing (I think) ocamlopt and then `-output-complete-exe`?

MisterDA replied:
Maybe it's just not necessary to do both and testing one of the bytecode or native is sufficient?

dra27 replied:
I think it's good to test both - although I can't remember if at the moment any of the headers behave differently between bytecode and native, both linking modes are "likely" for actual uses of C++ with OCaml.

dra27 commented on testsuite/tests/cxx-api/all_includes.ml:
```diff
@@ -1,13 +1,41 @@
- not-msvc;
- flags = "-ccopt -x -ccopt c++ -ccopt -std=c++11";
+ has-cxx;
+ not-clang;
```
Do neither has-cxx or the deps test not encompass this? (basically, I think it needs a comment!)

MisterDA replied:
This should filter out clang-cl, but I can do that in `deps.sh`.

dra27 commented on testsuite/tests/cxx-api/cxx-sarif.sh:
```diff
@@ -0,0 +1,85 @@
+  have_cygpath=false
+fi
+
+if [ "${ccomp_type:?}" = "cc" ]; then
```
`case` would be clearer here

dra27 commented on testsuite/tests/cxx-api/cxx-sarif.sh:
```diff
@@ -0,0 +1,85 @@
+  if "$have_cygpath"; then
+    ocamlsrcdir="$(cygpath -u "${ocamlsrcdir:?}")"
+    test_source_directory="$(cygpath -u "${test_source_directory:?}")"
+  fi
```
I don't _think_ this is necessary, certainly for Cygwin's mingw-w64 GCC and I don't think for MSYS2's either?

MisterDA replied:
I don't remember in what settings I needed this but at some point I definitely did. Let's remove them and see what happens.

dra27 commented on testsuite/tests/cxx-api/cxx-sarif.sh:
```diff
@@ -0,0 +1,85 @@
+    if [ -z "$length" ]; then
+      exit "$TEST_PASS"
+    fi
+  fi
```
Big comment here please as to what this is doing and why we skip that error. IIUC the point is that we get no warnings in the diagnostics apart from the ISO C++ one which we don't care about?

Is that string standardised, or is that sadly the best? From a distance, it's strange to have structured diagnostics, and then have to match a sentence!

MisterDA replied:
Unfortunately this is how GCC identifies this error:
```
{
  "ruleId": "-Werror=pedantic",
  "level": "error",
  "message": {
    "text": "ISO C++ forbids flexible array member 'dim'"
  },
}
```
It cannot be dissociated from other `-Wpedantic` errors apart from the error message. We could ignore all `-Wpendantic` errors, but we could miss interesting things too...
We may hope that the error message remains stable. I could also set `LC_LANG=C` to prevent error messages being localized.

dra27 commented on testsuite/tests/cxx-api/cxx-sarif.sh:
```diff
@@ -0,0 +1,85 @@
+| select(.ruleId
+        | contains("C4200") | not)
+| length
+EOF
```
Similarly big comment as above

dra27 commented on testsuite/tests/cxx-api/cxx-sarif.sh:
```diff
@@ -0,0 +1,85 @@
+    length=$(jq -f filter.jq stubs.sarif)
+    if [ -z "$length" ]; then
+      exit "$TEST_PASS"
+    fi
```
This is minor, but because one path is Unix and the other Windows I think worth doing - I'd factor this so that the result of this `case` block is to run the C++ compiler and create `filter.jq` - then put this as a common block below.

dra27 reviewed (commented):
Thanks for the work on this - a slightly noisy review, but C++ support for the headers is a recurring issue, and it's hard to test. We also have recurring issues with tests that analyse output directly (the debugger tests, etc.), so I definitely agree that analysing this with jq will make for a much less unpleasant experience when this test highlights a problem in future than if we had yet another awk script!!

What I am wondering, to put the proverbial cat amongst the pigeons, is if ocamltest is the right vehicle for this. I think the machinery for this would be simpler if there were a Makefile target (a la `test-in-prefix`) that just did this test if the support is available, especially given my suggestion about all-includes.h. I know that sounds like a heresy, but we're not _really_ testing the compiler itself in these tests? In some ways, it's almost like tools/checkstack?

gasche commented:
(Yes, I'm happy to have people step outside the ocamltest world if it makes a specific test substantially simpler/easier to write and to maintain.)

MisterDA commented:
Thank you both for the reviews, I have applied some of your suggestions, but not all yet. I haven't yet converted the test to use the Makefile infra.

MisterDA committed:
Revert "Skip sockaddr_cxx.ml in Jenkins"

This reverts commit 17a207b5ff49761eddc7c9f6ce76b11543f07a9c.

MisterDA committed:
Test for C++ API compatibility GCC and MSVC SARIF diagnostics

MisterDA committed:
Update Changes

MisterDA committed:
ocamltest: expose an has-cxx action, cxx and ccomp_type variables

MisterDA committed:
Disable the cxx-api test

MisterDA committed:
Fix sockaddr_cxx test under MSVC

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

MisterDA commented:
I've tried to move this test in to a Makefile and have the CI call it, as the `in-prefix` test, but I've found that it was not trivial, and that I had the feeling that I was duplicating features already found in ocamltest. I'm tempted to argue that if the test is working now, there's not much to gain from moving it out of ocamltest.
I've added more fixes to the PR, in particular I've used shell scripting instead of `find` and fixed invocations of MSVC, also in the sockaddr_cxx test.

gasche commented:
We discussed this today at the maintainer meeting. @dra27 is thinking of experimenting with a simpler way of doing this (keep the SARIF stuff, but without going through ocamltest). We agreed to wait a couple weeks to give him the opportunity to do so, but merge this PR if @dra27 does not end up having time to look at it.

dra27 committed:
Refactor all_includes.ml as a C++ API linking test

dra27 commented:
Running through [precheck#1113](https://ci.inria.fr/ocaml/job/precheck/1113/) - I'd suggest keeping the intermediate commits here, because it means that it'd concretely be in the history if/when we end up hitting a situation in future where we absolutely can't workaround the lack of granularity in `-Wpedantic`.

dra27 commented:
Rebased to take #14502 into account - I think this is good to merge?

gasche commented:
You decide! I was waiting for your final approval, so please feel free to go ahead and merge.

dra27 commented:
Sure - it becomes less clear when I’ve also edited it all!

gasche commented:
🙈

MisterDA commented:
I approve this PR on behalf of @MisterDA .
